packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp

```diff
@@ -0,0 +1,535 @@
+        Kokkos::View<int *, DeviceType> &ranks, Details::NearestPredicateTag,
+        Kokkos::View<double *, DeviceType> *distances_ptr = nullptr );
+
+    template <typename Query>
+    static void devise_strategy( Point epsilon,
+                                 Kokkos::View<Query *, DeviceType> queries,
+                                 BVH<DeviceType> const &bvh,
+                                 Kokkos::View<int *, DeviceType> &indices,
+                                 Kokkos::View<int *, DeviceType> &offset );
+
+    template <typename Query>
+    static void forward_queries( Teuchos::RCP<Teuchos::Comm<int> const> comm,
+                                 Kokkos::View<Query *, DeviceType> queries,
+                                 Kokkos::View<int *, DeviceType> indices,
+                                 Kokkos::View<int *, DeviceType> offset,
+                                 Kokkos::View<Query *, DeviceType> &fwd_queries,
+                                 Kokkos::View<int *, DeviceType> &fwd_ids,
+                                 Kokkos::View<int *, DeviceType> &fwd_ranks );
+
+    static void communicate_results_back(
+        Teuchos::RCP<Teuchos::Comm<int> const> comm,
+        Kokkos::View<int *, DeviceType> &indices,
+        Kokkos::View<int *, DeviceType> offset,
+        Kokkos::View<int *, DeviceType> &ranks,
+        Kokkos::View<int *, DeviceType> &ids,
+        Kokkos::View<double *, DeviceType> *distances_ptr = nullptr );
+
+    template <typename Query>
+    static void filter_results( Kokkos::View<Query *, DeviceType> queries,
+                                Kokkos::View<double *, DeviceType> distances,
+                                Kokkos::View<int *, DeviceType> &indices,
+                                Kokkos::View<int *, DeviceType> &offset,
+                                Kokkos::View<int *, DeviceType> &ranks );
+    static void
+    sort_results( Kokkos::View<int *, DeviceType> query_ids,
+                  Kokkos::View<int *, DeviceType> results,
+                  Kokkos::View<int *, DeviceType> ranks,
+                  Kokkos::View<double *, DeviceType> *distances_ptr = nullptr );
+
+    static void count_results( int n_queries,
+                               Kokkos::View<int *, DeviceType> query_ids,
+                               Kokkos::View<int *, DeviceType> &offset );
+
+    // NOTE: Would love to pass the distributor as a const reference but
+    // unfortunately the methods for executing the communication plan (e.g.
+    // doPostsAndWaits() in this case) are not declared with the const
+    // qualifier in Tpetra.
+    template <typename T>
+    static void send_across_network( Tpetra::Distributor &distributor,
+                                     Kokkos::View<T *, DeviceType> exports,
+                                     Kokkos::View<T *, DeviceType> imports );
+
+    static double epsilon;
+};
+
+template <typename DeviceType>
+double DistributedSearchTreeImpl<DeviceType>::epsilon = 1.0e-6;
```

Comment:
How did you pick `epsilon`? Is it independent of the mesh? Otherwise, we will get into trouble.